### DISCUSSION TITLE
rutorrent - don't hardcode endpoint (uri)

### DISCUSSION THREAD

**Discussion by al3xLvs** [linuxserver/Heimdall#752] [App issues]
My RPC url doesn't finish by `https://domain.tld/RPC2` but by `https://domain.tld/RPC`

We need to fix this code.
file: ruTorrent.php
```php
    public function url($endpoint)
    {
        return parent::normaliseurl($this->config->url).$endpoint;
    }

    public function getXMLRPCData($method)
    {
        ...
        $res = parent::execute($this->url('RPC2'), $this->attrs, $this->vars);
        ...
    }
```

**Comment by KodeStar** [comment]
v3 will work differently and URLs wont be hardcoded, the url will be automatically put in but you will be able to completely change the url if you want/need to. If you look at https://apps.heimdall.site/applications.json the sabnzbd one is in the new format that app.json will use to facilitate this. https://drive.google.com/file/d/11TIVFsKf4MA7r_3zJDvkkyvQklHPHYYr/view will give you an idea of how that will work

**Comment by Akashic101** [comment]
This still seems to be an issue. I'm running RuTorrent on a Seedbox meaning my URL looks like this `https://mole.website.eu/username/rutorrent/`. Entering that URL results in the error `There is an issue connecting to "https://mole.website.eu/username/rutorrent/RPC2". Please respect URL format "http(s)://IP:PORT". Unexpected response. Are credentials correct?`

**Reply by mwsmws22** [reply]
Same issue here. Did you happen to find a solution?

**Reply by Akashic101** [reply]
Sadly not, the issue seems to persist to this day

**Reply by mwsmws22** [reply]
That's unfortunate... thanks for the confirmation

**Comment by mwsmws22** [comment]
In regards to the issue encountered by @Akashic101 and myself when using a seedbox, I have a workaround:

1. For the url, use the following in the config section `https://insert.server.here.me/rutorrent/plugins/httprpc/action.php`. Replace the server with whatever yours is.
2. In `heimdall/config/www/SupportedApps/ruTorrent/ruTorrent.php` edit the following line:
```diff
try {
    $res = parent::execute(
-       $this->url("RPC2"),
+       $this->config->url,
        $this->attrs,
        $this->vars
    );
} catch (\GuzzleHttp\Exception\RequestException $e) {
    return ""; // Connection failed, display default response
}
```
3. In `heimdall/config/www/SupportedApps/ruTorrent/config.blade.php` edit the following line:
```diff
<div class="input">
     <label>{{ __('app.apps.password') }}</label>
-    {!! Form::text('config[private_apikey]', isset($item) ? $item->getconfig()->private_apikey : null, ['placeholder' => __('app.apps.apikey'), 'id' => 'private_apikey', 'class' => 'form-control config-item']) !!}
+    {!! Form::input('password', 'config[password]', '', ['placeholder' => __('app.apps.password'), 'data-config' => 'password', 'class' => 'form-control config-item']) !!}
</div>
```
That's it, just make sure your username/password are correct.

**Reply by Akashic101** [reply]
That did the trick, thank you very much for your help

**Reply by mwsmws22** [reply]
No problem!
